// astro-tunnel/dist/index.js
// src/index.ts
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { startTunnel } from "untun";
function createAstroTunnelIntegration(options = {
  port: 4321,
  hostname: "localhost",
  protocol: "http",
  verifyTLS: false,
  acceptCloudflareNotice: false
}) {
  return {
    name: "astro-tunnel",
    hooks: {
      "astro:config:setup": ({ addDevToolbarApp }) => {
        const __filename = fileURLToPath(import.meta.url);
        const __dirname = dirname(__filename);
        addDevToolbarApp(join(__dirname, "./astro-tunnel.js"));
      },
      "astro:server:setup": ({ server }) => {
        let tunnel;
        server.hot.on(
          "astro-dev-toolbar:astro-tunnel:initialized",
          async () => {
            server.hot.send("astro-tunnel:tunnel-url", {
              url: await tunnel?.getURL()
            });
          }
        );
        server.hot.on("astro-dev-toolbar:astro-tunnel:toggled", async () => {
          server.hot.send("astro-tunnel:tunnel-url", {
            url: await tunnel?.getURL()
          });
        });
        server.hot.on(
          "astro-tunnel:toggled",
          async (data) => {
            if (data.checked) {
              tunnel = await startTunnel(options);
              server.hot.send("astro-tunnel:tunnel-url", {
                url: await tunnel?.getURL()
              });
            } else {
              await tunnel?.close();
              server.hot.send("astro-tunnel:tunnel-url", {
                url: void 0
              });
              tunnel = void 0;
            }
          }
        );
      }
    }
  };
}
export {
  createAstroTunnelIntegration as default
};
